cloud-aws-cloudformation/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/upgrade/operation/AwsRdsUpgradeValidatorService.java

```diff
@@ -0,0 +1,80 @@
+import com.sequenceiq.cloudbreak.common.database.Version;
+import com.sequenceiq.cloudbreak.util.MajorVersionComparator;
+
+@Service
+public class AwsRdsUpgradeValidatorService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AwsRdsUpgradeValidatorService.class);
+
+    public void validateRdsCanBeUpgraded(RdsInfo rdsInfo, Version targetMajorVersion) {
+        LOGGER.debug("Rds state is: {} and db instance statuses are: {} ", rdsInfo.getRdsState(), rdsInfo.getDbArnToInstanceStatuses());
+
+        validateRdsIsAvailableOrUpgrading(rdsInfo);
+        validateRdsMajorVersionIsSmallerThanTarget(rdsInfo, targetMajorVersion);
+    }
+
+    public void validateUpgradePresentForTargetMajorVersion(Set<RdsEngineVersion> upgradeTargetsForMajorVersion) {
+        if (upgradeTargetsForMajorVersion.isEmpty()) {
+            String message = "There are no matching RDS upgrade versions to choose from.";
+            LOGGER.warn(message);
+            throw new CloudConnectorException(message);
+        }
+    }
+
+    public void validateClusterHasASingleVersion(Set<String> dbVersions) {
+        if (dbVersions.isEmpty()) {
+            String message = "Current DB version could not be obtained";
+            LOGGER.warn(message);
+            throw new CloudConnectorException(message);
+        }
+
+        if (dbVersions.size() > 1) {
+            String message = String.format("RDS is on multiple versions (%s), cannot proceed with its upgrade", dbVersions);
+            LOGGER.warn(message);
+            throw new CloudConnectorException(message);
+        }
+    }
+
+    private void validateRdsIsAvailableOrUpgrading(RdsInfo rdsInfo) {
+        if (OTHER == rdsInfo.getRdsState() || UNKNOWN == rdsInfo.getRdsState()) {
+            String message = String.format("RDS upgrade is not possible as one or more instances are not in a correct state: %s",
+                    getNotApplicableStates(rdsInfo.getDbArnToInstanceStatuses()));
+            LOGGER.warn(message);
+            throw new CloudConnectorException(message);
+        }
+    }
+
+    private void validateRdsMajorVersionIsSmallerThanTarget(RdsInfo rdsInfo, Version targetMajorVersion) {
+        if (new MajorVersionComparator().compare(rdsInfo.getRdsEngineVersion(), new RdsEngineVersion(targetMajorVersion.getMajorVersion())) >= 0) {
+            String message = String.format("Cannot start upgrade as the RDS major version %s is not smaller than the target version %s",
+                    rdsInfo.getRdsEngineVersion(), targetMajorVersion);
+            LOGGER.debug(message);
+            throw new CloudConnectorException(message);
+        }
+    }
+
+    private Set<String> getNotApplicableStates(Map<String, String> dbArnToInstanceStatuses) {
+        return dbArnToInstanceStatuses.entrySet().stream()
+                .filter(entry -> !AVAILABLE.getState().equals(entry.getValue())
+                        || !UPGRADING.getState().equals(entry.getValue()))
```

Comment:
Shouldn't it be an AND between the operands? if the status is UPGRADING the filter would return true now.

Reply:
Yes, of course. Thank you very much for catching this.